spatial/curve/hilbert.go

```diff
@@ -0,0 +1,241 @@
+// Copyright ©2024 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package curve
+
+// Hilbert2D is a 2-dimensional Hilbert curve.
+type Hilbert2D struct{ Order int }
+
+// Dims returns the spatial dimensions of the curve, which is {2ᵏ, 2ᵏ}, where k
+// is the order.
+func (h Hilbert2D) Dims() []int { return []int{1 << h.Order, 1 << h.Order} }
+
+// Len returns the length of the curve, which is 2ⁿᵏ, where n is the dimension
+// (2) and k is the order.
+//
+// Len will overflow on a 32-bit architecture if the order is ≥ 16.
+func (h Hilbert2D) Len() int { return 1 << (2 * h.Order) }
+
+func (h Hilbert2D) rot(n int, v []int, d int) {
+	switch d {
+	case 0:
+		swap{0, 1}.do(n, v)
+	case 3:
+		flip{0, 1}.do(n, v)
+	}
+}
+
+// Pos returns the linear position of the spatial coordinate along the curve.
+// Pos modifies v.
+//
+// Pos will overflow on a 32-bit architecture if the order is ≥ 16.
+func (h Hilbert2D) Pos(v []int) int {
+	var d int
+	for n := h.Order - 1; n >= 0; n-- {
+		rx := (v[0] >> n) & 1
+		ry := (v[1] >> n) & 1
+		rd := ry<<1 | (ry ^ rx)
+		d += rd << (2 * n)
+		h.rot(h.Order, v, rd)
+	}
+	return d
+}
+
+// WriteCoordTo writes the spatial coordinates of pos to v.
```

Comment:
Note that `len(v) >= 2` otherwise `WriteCoordTo` will panic. Consider making this strict (i.e. `len(v) == 2`) to be consistent with other length requirement check that we impose elsewhere.

Reply:
Done